ha/core/config/consul_kv_store.py

```diff
@@ -0,0 +1,185 @@
+# with this program. If not, see <https://www.gnu.org/licenses/>. For any questions
+# about this software or licensing, please email opensource@seagate.com or
+# cortx-questions@seagate.com.
+
+import consul
+import errno
+import socket
+from urllib.parse import urlparse
+
+class ConsulKvStore:
+    """ Represents a Consul kv Store """
+
+    name = "consul"
+    version = "/v1"
+
+    def __init__(self, url: str):
+        """
+        Consul KV store.
+
+        Args:
+            url (str): Consul url with prefix
+
+        Example:
+            ConsulKvStore("consul://localhost:8500/cortx/ha/system/config")
+            ConsulKvStore("consul:///v1/cortx/ha/system/config")
+        Here localhost:8500 is connection and cortx/ha/system/config is prefix.
+        """
+        url_spec = urlparse(url)
+        store_type = url_spec.scheme
+        store_loc = url_spec.netloc
+        store_path = url_spec.path
+        self._prefix: str = None
+        self._verify(store_type=store_type, store_loc=store_loc, store_path=store_path)
+        self._consul = self._get_connection(store_loc=store_loc, store_path=store_path)
+        self._consul.kv.put(self._prepare_key(""), None)
+
+    def _verify(self, store_type: str, store_loc: str, store_path: str):
+        """
+        Verify connection detail.
+
+        Args:
+            store_type (str): Type of connection or DB name.
+            store_loc (str): Connection detail like host and port.
+            store_path (str): Prefix for connection.
+        """
+        if store_type != ConsulKvStore.name:
+            raise Exception(f"Invalid storage type {store_type}. Correct one is {ConsulKvStore.name}")
+        if store_path == "":
+            raise Exception("Invalid prefix. It cannot be empty.")
+        if store_loc != "":
+            try:
+                socket.gethostbyname(store_loc.split(":")[0])
+                int(store_loc.split(":")[1])
+            except Exception as e:
+                raise Exception(f"Invalid host or port name. Refused connection to {store_loc}. Error: {e}")
+
+    def _get_connection(self, store_loc: str, store_path: str):
+        """
+        Prepare consul connection.
+
+        Args:
+            store_type (str): Type of connection or DB name.
+            store_loc (str): Connection detail like host and port.
+            store_path (str): Prefix for connection.
+
+        Return:
+            Object: Consul object.
+        """
+        self._prefix = ConsulKvStore.version + store_path
+        if store_loc == "":
+            return consul.Consul()
+        host: str = store_loc.split(":")[0]
+        port: int = int(store_loc.split(":")[1])
+        return consul.Consul(host=host, port=port)
+
+    def _verify_data(self, *args):
+        """
+        Verify data.
+        """
+        for data in args:
+            if data is None or data == "":
+                raise Exception("Invalid parameter. Data cannot be none or empty.")
```

Comment:
Why data can't be empty? Consul doesn't support it? What if I want to create keys with an empty value?

Reply:
yes data can be empty 
modified code to handle empty data